modules/core/src/main/java/org/apache/ignite/internal/ThinProtocolFeature.java

```diff
@@ -0,0 +1,70 @@
+package org.apache.ignite.internal;
+
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.EnumSet;
+
+/**
+ * The base feature class.
+ */
+public interface ThinProtocolFeature {
+    /**
+     * @return Feature ID.
+     */
+    int featureId();
+
+    /**
+     * @return Feature's name.
+     */
+    String name();
+
+    /**
+     * @param features Features set.
+     * @return Byte array representing all supported features by current node.
+     * @param <E> Features type (JDBC, thin cli, ODBC)
+     */
+    static <E extends Enum<E> & ThinProtocolFeature> byte[] featuresAsBytes(E[] features) {
+        final BitSet set = new BitSet();
+
+        for (ThinProtocolFeature f : features)
+            set.set(f.featureId());
+
+        return set.toByteArray();
+    }
+
+    /**
+     * @param features Features set.
+     * @return Byte array representing all supported features.
+     * @param <E> Features type (JDBC, thin cli, ODBC)
+     */
+    static <E extends Enum<E> & ThinProtocolFeature> byte[] featuresAsBytes(Collection<E> features) {
+        final BitSet set = new BitSet();
+
+        for (ThinProtocolFeature f : features)
+            set.set(f.featureId());
+
+        return set.toByteArray();
+    }
+
+    /**
+     * Create EnumSet of supported features encoded by the bytes array.
+     *
+     * @param <E> Features type (JDBC, thin cli, ODBC)
+     * @param in Byte array representing all supported features.
+     */
+    static <E extends Enum<E> & ThinProtocolFeature> EnumSet<E> enumSet(byte [] in, E[] values) {
```

Comment:
I would change second parameter to `Class<E> cls`. It should make API less error prone.

```
static <E extends Enum<E> & ThinProtocolFeature> EnumSet<E> enumSet(byte [] in, Class<E> cls) {
    ...
    EnumSet<E> set = EnumSet.noneOf(cls);
    ...
    for (E e : cls.getEnumConstants()) {
    ...
```